src/components/queryExpressionEditor/formMode/FormMode.tsx

```diff
@@ -0,0 +1,142 @@
+import * as React from 'react';
+import * as _ from 'underscore';
+import {convertUndefinedAndNullToEmptyString} from '../../../utils/FalsyValuesUtils';
+import {BOOLEAN_OPRATOR_SELECT_ID} from '../booleanOperatorSelect/BooleanOperatorSelect';
+import {ExpressionEditorConnected} from '../expressionEditor/ExpressionEditorConnected';
+import {IExpressionEditorState} from '../expressionEditor/ExpressionEditorReducers';
+import {QueryTrigger} from '../queryTrigger/QueryTrigger';
+import {IField} from '../responseParser/ResponseParser';
+import * as styles from './FormMode.scss';
+
+export const EXPRESSION_EDITOR_ID: string = 'expression-editor';
+
+export interface IFormModeOwnProps {
+    queryTrigger: QueryTrigger;
+    updateQueryExpression: (expression: string) => void;
+}
+
+export interface IFormModeOwnState {
+    fields?: IField[];
+    expressionEditors?: JSX.Element[];
+}
+
+export interface IFormModeStateProps {
+    expressionEditorsState?: IExpressionEditorState[];
+}
+
+export interface IFormModeDispatchProps {
+    selectListBoxOption?: (id: string, multi: boolean, value: string) => void;
+}
+
+export interface IFormModeProps extends IFormModeOwnProps, IFormModeStateProps, IFormModeDispatchProps {}
+
+export class FormMode extends React.Component<IFormModeProps, IFormModeOwnState> {
+    private expressionEditorNewId: number;
+
+    constructor(props: IFormModeOwnProps) {
+        super(props);
+        this.state = {fields: [], expressionEditors: []};
+        this.expressionEditorNewId = 0;
+    }
+
+    componentWillReceiveProps(nextProps: IFormModeProps) {
+        const finaleExpression = this.getFinalExpression(nextProps);
+        this.props.updateQueryExpression(finaleExpression);
+    }
+
+    async componentDidMount() {
+        const getfields = await this.props.queryTrigger.getFields();
+        this.setState({fields: getfields});
+        this.addExpressionEditor();
+    }
+
+    private addExpressionEditor() {
+        const newSate = this.state.expressionEditors;
+        newSate.push(
+            <ExpressionEditorConnected
+                key={`${EXPRESSION_EDITOR_ID}-${this.expressionEditorNewId}`}
+                id={`${EXPRESSION_EDITOR_ID}-${this.expressionEditorNewId}`}
+                fields={this.state.fields}
+                queryTrigger={this.props.queryTrigger}
+                addExpressionEditor={() => this.addExpressionEditor()}
+                deleteExpressionEditor={(id: string) => this.deleteExpressionEditor(id)}
+            />,
+        );
+        this.setState({expressionEditors: newSate});
+        this.expressionEditorNewId++;

```

Comment:
i would have a getNextExpressionEditor method, and since the next expressionEditor id is related to this.state.expressionEditors.length , i wouldn't bother managing this reference and simply remove it.   
  
```
    private addExpressionEditor() {
        this.setState({expressionEditors: [...this.state.expressionEditors, this.getNextExpressionEditor()]});
    }

    private getNextExpressionEditor() {
        return (
            <ExpressionEditorConnected
                key={`${EXPRESSION_EDITOR_ID}-${this.state.expressionEditors.length}`}
                id={`${EXPRESSION_EDITOR_ID}-${this.state.expressionEditors.length}`}
                fields={this.state.fields}
                queryTrigger={this.props.queryTrigger}
                addExpressionEditor={() => this.addExpressionEditor()}
                deleteExpressionEditor={(id: string) => this.deleteExpressionEditor(id)}
            />
        );
    }  
```